### ISSUE TITLE
mosquitto_loop crash on iOS at line FD_SET(mosq->sockpairR, &readfds);

### ISSUE THREAD

**Issue by haok3166** [eclipse/mosquitto#327]
Hi there,

I am using the latest mosquitto lib on iOS to receive streaming data from mqtt servers. It works perfectly except there are some crashes in mosquitto_loop method. However, I could not reproduce this crash on local tests. I saw the crash report from crashlytics which has around 0.5% crash rate.

			if(mosq->sockpairR != INVALID_SOCKET && FD_ISSET(mosq->sockpairR, &readfds)){
#ifndef WIN32
				if(read(mosq->sockpairR, &pairbuf, 1) == 0){
				}
#else
				recv(mosq->sockpairR, &pairbuf, 1, 0);
#endif
				/* Fake write possible, to stimulate output write even though
				 * we didn't ask for it, because at that point the publish or
				 * other command wasn't present. */
				FD_SET(mosq->sock, &writefds);
			}

The crash always happens on line 963 FD_SET(mosq->sock, &writefds);, I have tried to comment out this line and the app has no crash on our test iOS devices. I am wondering if there is a side effect doing this? Will this change cause crash once it is released to App Store?

I really apprecite your help and thank you for creating such great MQTT product.

Best Regards

**Comment by ralight** [comment]
The sockpairR/W sockets are there to ensure if you call e.g. mosquitto_publish() when you're in threaded operation and the other thread is stuck in select(), that it can break out of select() before the timeout.

The call for FD_SET() is modifying the output of select() so that the subsequent code thinks that mosq->sock was actually writable - so not having to wait until the next call to select. There are two reasons I can think of that would cause this - either mosq->sock is too big - which it could be if you're opening lots of connections in one app (most likely >1024). More likely, if mosq->sock has been closed just before this call and set to -1, then FD_SET(-1, ...) would certainly crash.

The obvious solution is to check the validity of mosq->sock just before the call to FD_SET, so

    if(mosq->sock >= 0) FD_SET(...);

Would it be possible to test that to confirm the hypothesis?

Ultimately, commenting out the call to FD_SET() will at its worst just potentially introduce some latency into outgoing messages, so it may not be the end of the world for you.

**Comment by haok3166** [comment]
Hi Roger,

Thank you very much for your reply.

I did think of the possible reason that the connection limitation (1024) was reached on iOS devices, and I saw from crashlytics that the devices crashed had low free memory. I think it is one of the possible reasons. However, the app only calls a few REST APIs and create at most 5 mqtt connections. So technically, the chance reaching the limit is pretty low. I believe your suggestion to check mosq->sock before FD_SET will work. I will try to add if(mosq->sock >= 0) FD_SET(...); to see if it can fix this issue.

I have tried lots of MQTT implementations, the mosquitto is so far the most efficient and stable implementation based on my tests. Another question, it seems the broker will disconnect the publisher if the publisher publish too many message (qos 0), is there a way to increase the broker's buffer size so that the broker can deliver messages later instead of disconnecting the publisher (lose messages)?

Thanks again for creating this great product.

Regards
